File: 백준/단계 별로 풀어보기/강한_연결_요소/2150_Strongly_Connected_Component.py
import sys
input = sys.stdin.readline
IP = lambda: map(int, sys.stdin.readline().split())
sys.setrecursionlimit(10 ** 9)

def Kosaraju():
    def forward_dfs(present):
        visited[present] = True
        for next_vertex in forward_edge[present]:
            if not visited[next_vertex]: forward_dfs(next_vertex)
        stack.append(present)
    
    def backward_dfs(present):
        visited[present] = True
        part_scc_set.append(present)
        for next_vertex in backward_edge[present]:
            if not visited[next_vertex]: backward_dfs(next_vertex)
    
    
    V, E = IP()
    forward_edge = [[] for i in range(V + 1)]
    backward_edge = [[] for i in range(V + 1)]
    visited = [False] * (V + 1)
    stack = []; final_scc_set = []
    
    for i in range(E):
        a, b = IP()
        forward_edge[a].append(b)
        backward_edge[b].append(a)
    
    
    for i in range(1, V + 1): forward_edge[i].sort(); backward_edge[i].sort()
    
    for i in range(1, V + 1):
        if not visited[i]: forward_dfs(i)
    
    visited = [False] * (V + 1)
    while stack:
        p = stack.pop()
        if visited[p]: continue
        part_scc_set = []
        backward_dfs(p)
        final_scc_set.append(part_scc_set[:])
    
    for each in final_scc_set: each.sort()
    final_scc_set.sort()
    print(len(final_scc_set))
    for each in final_scc_set: print(' '.join(map(str, each)), end = ' -1\n')
    
    
    
def Tarjan():
    def dfs(present):
        stack.append(present)
        global n
        dfsn[present] = upper[present] = n
        n += 1
        for next_vertex in edge[present]:
            #교차 간선
            if finished[next_vertex]: continue
            #단순 진행
            elif dfsn[next_vertex] < 0:
                temp_up = dfs(next_vertex)
                if temp_up < upper[present]: upper[present] = temp_up
            #역방향 간선
            elif upper[next_vertex] < upper[present]: upper[present] = upper[next_vertex]
        
        #자신으로부터 조상노드로 갈 수 없다면 scc를 추출한다
        if upper[present] == dfsn[present]:
            p = None
            part_scc = []
            while p != present:
                p = stack.pop()
                finished[p] = True
                part_scc.append(p)
            final_scc_set.append(sorted(part_scc))
        
        return upper[present]
    
    
    V, E = IP()
    edge = [[] for i in range(V + 1)]
    dfsn = [-1] * (V + 1)
    upper = [0] * (V + 1)
    finished = [False] * (V + 1)
    stack = []; final_scc_set = []
    global n
    n = 1
    
    for _ in range(E):
        a, b = IP()
        edge[a].append(b)
    
    for i in range(1, V + 1):
        if not finished[i]: dfs(i)
    
    print(len(final_scc_set))
    for each in sorted(final_scc_set): sys.stdout.write(' '.join(map(str, each)) + ' -1\n')
    

if "__main__" == __name__:
    #Kosaraju()
    Tarjan()

File: 백준/단계 별로 풀어보기/강한_연결_요소/2150_Strongly_Connected_Component.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
using namespace std;

int dfs(int present, int &n, vector<int>& stack, vector<int>& dfsn, vector<int>& upper, vector<bool>& finished, vector< vector<int> >& final_scc_set, vector< vector<int> >& edge)
{
    stack.push_back(present);
    dfsn[present] = n;
    upper[present] = n;
    n++;
    
    for (auto next_vertex : edge[present]){
        if (finished[next_vertex]) continue;
        else if (dfsn[next_vertex] < 0){
            int temp_up = dfs(next_vertex, n, stack, dfsn, upper, finished, final_scc_set, edge);
            upper[present] = min(temp_up, upper[present]);
        }
        else upper[present] = min(upper[present], upper[next_vertex]);
    }
    
    if (upper[present] == dfsn[present]){
        int p = -1;
        vector<int> part_scc;
        
        while (p != present){
            p = stack.back();
            stack.pop_back();
            finished[p] = true;
            part_scc.push_back(p);
        }
        sort(part_scc.begin(), part_scc.end());
        final_scc_set.push_back(part_scc);
    }
    return upper[present];
}

int main()
{
	cin.tie(NULL);
	cout.tie(NULL);
	ios_base::sync_with_stdio(false);
	
	
	//입력부
	int V, E, a, b;
	cin >> V >> E;
	
	vector< vector<int> > edge(V + 1, vector<int>(0, 0));
	for (int i = 0; i < E; i++){
	    cin >> a >> b;
	    edge[a].push_back(b);
	}
	
	//dfs 전처리
	vector<int> dfsn(V + 1, -1);
	vector<int> upper(V + 1, 0);
	vector<bool> finished(V + 1, false);
	vector<int> stack;
	vector< vector<int> > final_scc_set;
	int n = 1;
	
	for (int i = 1; i <= V; i++){
	    if (!finished[i]) dfs(i, n, stack, dfsn, upper, finished, final_scc_set, edge);
	}
	
	cout << final_scc_set.size() << '\n';

	sort(final_scc_set.begin(), final_scc_set.end());
	for (auto each : final_scc_set){
	    for (auto i : each) cout << i << " ";
	    cout << "-1\n";
	}
}